### PR TITLE
Feat/stripe payment

### PR DESCRIPTION
**Pull request by zweroboy1**
## Describe your changes
- Integrate application with Stripe.
- Add order creation on payment success.
- Add filters for best vendors. 

## Issue ticket code (and/or) and link
- [Link to JIRA ticket](https://2024internship.atlassian.net/browse/SCRUM-201)
  
### **General**
- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [x] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [x] Date and time formats are on the constants
- [x] Functions are public only if it's used outside the class
- [x] No hardcoded values
- [ ] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Backend
- [x] Swagger documentation updated
- [x] Database requests are optimized and not redundant
- [ ] Unit tests written
- [x] use ConfigService instead of process.env
- [x] use transactions if there is a call chain that mutates data in different tables
- [x] use @index decorator for frequently requested data

### CODE REVIEW

File: src/modules/stripe/stripe.service.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+      const productsTotalPrice = products.reduce(
+        (sum, product) => sum + Number(product.price),
+        0,
+      );
+
+      if (
+        Math.abs(productsTotalPrice + shippingPrice - total) > minDifference
+      ) {
+        throw new ConflictException(Errors.PRICES_DO_NOT_MATCH);
```
Comment (lyudad) [inline src/modules/stripe/stripe.service.ts:72]: how you will handle it on frontend?

File: src/modules/stripe/stripe.service.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    signature: string,
+  ): Promise<Stripe.Event> {
+    try {
+      const consrtuctedEvent = this.StripeApi.webhooks.constructEvent(
+        event,
+        signature,
+        this.configService.get<string>('STRIPE_WEBHOOK_SECRET'),
+      );
+
+      console.log(consrtuctedEvent);
```
Comment (lyudad) [inline src/modules/stripe/stripe.service.ts]: ?

File: src/modules/stripe/stripe.service.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        (order) => order.transferGroup === transferGroup,
+      ).userId;
+
+      await this.cartService.emptyCart(userId);
+    }
+
+    return { received: true };
+  }
+
+  async checkSignature(
```
Comment (lyudad) [inline src/modules/stripe/stripe.service.ts:160]: are you sure that it should be open route

File: src/modules/stripe/stripe.service.ts (hunk 1 above)
Comment (zweroboy1) [reply src/modules/stripe/stripe.service.ts:72]: The user will see a toast notification indicating that the prices of the items have changed. Then, the cart will reload, showing the new prices, and the user will be able to confirm the payment.

File: src/modules/stripe/stripe.service.ts (hunk 3 above)
Comment (zweroboy1) [reply src/modules/stripe/stripe.service.ts:160]: The webhook route is used by Stripe to transmit transaction data. It is public, but the data is verified using a secret key.

File: src/modules/stripe/stripe.service.ts (hunk 2 above)
Comment (zweroboy1) [reply src/modules/stripe/stripe.service.ts]: This was needed to see in the logs of the deployed version what exact data Stripe sends after a payment is made (and whether this data exists at all). I will remove this.

**Review by lyudad — APPROVED** [review]

**Pull request merged**